## External Validation of LASSO Model

**Dataset:** GSE42568: 104 breast cancer + 17 normal breast tissue transcriptomic profiles (Affymetrix HG-U133 Plus 2.0, GPL570).

**Objective** Utilize the saved models from analysis.ipynb and recreate a similar environment with an external dataset for validation of the model.

In [1]:
import os
import numpy as np
import pandas as pd
import joblib

#accuracy alone can be misleading, we'd want to add classification reports, conf_matrix
#as well as Receiving Operator Characteristic - Area Under the Curve (ROC_AUC)
#ROC_AUC is used to determine how well a binary classification model (which this is)
#distinguishes between two classes regardless of classification thresholds (which is 0.5 here)
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix,
    roc_auc_score,
)

EXTERNAL_DATA_PATH = "data/GSE42568_series_matrix.txt"
EXPORT_DIR         = "exported_model"


## Data Loading

A similar process to earlier

In [2]:
#load exported assets, scaler, model, feature list
scaler   = joblib.load(os.path.join(EXPORT_DIR, "training_scaler.joblib"))
model    = joblib.load(os.path.join(EXPORT_DIR, "lasso_1se_model.joblib"))
features = joblib.load(os.path.join(EXPORT_DIR, "retained_features.joblib"))

print("  LOADED TRAINING ARTIFACTS")
print(f"  Model type         : {type(model).__name__}")
print(f"  Model alpha        : {model.alpha:.6f}")
print(f"  # retained probes  : {len(features)}")
print(f"  Scaler n_features  : {scaler.n_features_in_}")
print(f"  Retained features  : {features}")

  LOADED TRAINING ARTIFACTS
  Model type         : Lasso
  Model alpha        : 0.129748
  # retained probes  : 15
  Scaler n_features  : 11142
  Retained features  : ['201650_at', '202014_at', '202988_s_at', '203549_s_at', '205478_at', '206243_at', '209550_at', '209686_at', '210964_s_at', '212112_s_at', '218124_at', '219398_at', '219738_s_at', '221009_s_at', '266_s_at']


In [3]:
#Read GEO file and extract labels, similar process to earlier, with a different parsing method for target 
#(since they use a different format)

y_test = None
sample_ids = None
matrix_start_line = None

with open(EXTERNAL_DATA_PATH, "r") as f:
    for line_num, line in enumerate(f, start=1):
        #this one uses tissure:normal, tissure:cancer
        if line.startswith("!Sample_characteristics_ch1") and "tissue:" in line:
            fields = line.strip().split("\t")[1:]
            y_test = np.array(
                [0 if "normal breast" in v else 1 for v in fields],
                dtype=int,
            )
        if line.startswith("!Sample_geo_accession"):
            sample_ids = [
                v.strip().strip('"') for v in line.strip().split("\t")[1:]
            ]
        if line.startswith("!series_matrix_table_begin"):
            matrix_start_line = line_num
            break

n_normal = int((y_test == 0).sum())
n_tumor  = int((y_test == 1).sum())
print(f"External dataset labels parsed:  {len(y_test)} samples  "
      f"({n_normal} normal, {n_tumor} tumor)")

External dataset labels parsed:  121 samples  (17 normal, 104 tumor)


In [4]:
#load it as a DF
expr = pd.read_csv(
    EXTERNAL_DATA_PATH,
    sep="\t",
    skiprows=matrix_start_line,
    index_col=0,
)

if expr.index[-1].startswith("!"):
    expr = expr.iloc[:-1]

#transpose
X_test_df = expr.T
X_test_df = X_test_df.apply(pd.to_numeric, errors="coerce").fillna(0)

print(f"External expression matrix loaded: {X_test_df.shape}  (samples × probes)")

External expression matrix loaded: (121, 54675)  (samples × probes)


## LOG2 SCALE GUARD

The training set (GSE15852) is raw MAS5-style intensity (max ~19 000), so `analysis.ipynb` applied `log2(x+1)` before everything else.
The external set has to be on the same scale before we compare anything, so we apply the **same guard** used in training:
if the max value is well above ~50 the data is raw and gets `log2(x+1)`, otherwise it is already log2 and is left alone.

GSE42568 is distributed as RMA-normalised data, which is already log2 (max ~16), so we expect the guard to leave it untouched.
We print min / median / max before and after so the check is visible.

In [5]:
def _scale_summary(df):
    vals = df.values
    return f"min = {vals.min():.2f}   median = {np.median(vals):.2f}   max = {vals.max():.2f}"

print("LOG2 SCALE GUARD")
print(f"Before : {_scale_summary(X_test_df)}")

#same rule as the training notebook
max_val = X_test_df.max().max()
if max_val > 50:
    X_test_df = np.log2(X_test_df + 1)
    print(f"Max value {max_val:.2f} > 50, raw intensities detected, log2(x+1) applied.")
else:
    print(f"Max value {max_val:.2f} (<= 50), data is already log2 scale (RMA), no transform applied.")

print(f"After  : {_scale_summary(X_test_df)}")

LOG2 SCALE GUARD
Before : min = 2.31   median = 2.84   max = 16.05
Max value 16.05 (<= 50), data is already log2 scale (RMA), no transform applied.


After  : min = 2.31   median = 2.84   max = 16.05


## Feature Alignment

The external dataset likely has different set of PROBE_IDs than the one we trained on.

We need to:
1. Keep ONLY the columns matching retained features
2. ZERO fill any retained features that did not exist in dataset (if performance is still good after this, it suggests model is doing very well)
3. Reorder columns to match training feature order

**Important**: we must provide the FULL scaler feature set, not just the
retained ones.  The training scaler was fit on ALL variance-filtered
features (11,142).  We first need to build a matrix with ALL those
columns (filling missing ones with 0) so that .transform() works.
Then we subset to the retained features AFTER scaling.

In [6]:
external_probes = set(X_test_df.columns)
retained_set    = set(features)

present  = retained_set & external_probes
missing  = retained_set - external_probes

print(f"Feature alignment:")
print(f"  Retained features     : {len(features)}")
print(f"  Found in external data: {len(present)}")
print(f"  Missing (zero-filled) : {len(missing)}")
if missing:
    print(f"    → {sorted(missing)}")

Feature alignment:
  Retained features     : 15
  Found in external data: 15
  Missing (zero-filled) : 0


In [7]:
scaler_features = list(scaler.feature_names_in_) if hasattr(scaler, "feature_names_in_") else None

if scaler_features is not None:
    # build a DataFrame aligned to ALL scaler features, in the scaler's column order.
    # reindex fills probes missing from the external data with 0 in one step
    # (much faster than assigning ~11k columns one at a time).
    X_test_full = X_test_df.reindex(columns=scaler_features, fill_value=0.0)

    common_full = set(scaler_features) & external_probes
    print(f"Full scaler alignment : {len(scaler_features)} columns "
          f"({len(common_full)} found, {len(scaler_features) - len(common_full)} zero-filled)")
    print()

    # we use .transform() — NOT .fit_transform() — so the centering and
    # scaling parameters come exclusively from the training data.
    X_test_scaled_full = scaler.transform(X_test_full)

print(f"Scaled test matrix shape: {X_test_scaled_full.shape}")
print()

Full scaler alignment : 11142 columns (11136 found, 6 zero-filled)



Scaled test matrix shape: (121, 11142)



## Evaluate

A Lasso model outputs continuous predictions (roughly [0,1], but not bounded).
Threshold is set at 0.5 as standard binary classification does.
This can be adjusted at your own discretion though.

This Dataset utilized the platform GPL570, according to my research online, this should be fine since GPL570 is simply a superset of GPL96.

**Why two options?** The log guard puts both datasets on a log2 scale, but they are still normalised differently
(GSE15852 = MAS5, GSE42568 = RMA). RMA values sit on a different baseline, so a given probe can be several log2 units
higher or lower in the external data for purely technical reasons. The training scaler's means/SDs come from the MAS5 data,
so applying it to RMA data shifts every standardised value. ROC-AUC (rank based) can survive this, the 0.5 threshold cannot.

We therefore evaluate two ways and report them side by side:
- **Option A (primary):** per-dataset standardisation of the 15 panel probes.
- **Option B:** training scaler on the log-aligned data (the original approach).

Both use the exact same frozen LASSO coefficients and intercept. Nothing is refit on external data, and the external labels are only used for the final metrics below.

In [8]:
#shared evaluation helper so both options are reported identically
def evaluate(y_true, y_scores, name):
    y_pred = (y_scores > 0.5).astype(int)
    cm = confusion_matrix(y_true, y_pred, labels=[0, 1])

    # derive Sensitivity and Specificity from the confusion matrix
    TN, FP, FN, TP = cm.ravel()
    res = {
        "Accuracy"   : accuracy_score(y_true, y_pred),
        "ROC-AUC"    : roc_auc_score(y_true, y_scores),
        "Sensitivity": TP / (TP + FN) if (TP + FN) > 0 else 0.0,   # recall for Tumor
        "Specificity": TN / (TN + FP) if (TN + FP) > 0 else 0.0,   # recall for Normal
        "TN": TN, "FP": FP, "FN": FN, "TP": TP,
        "Normal score mean"  : y_scores[y_true == 0].mean(),
        "Normal score median": np.median(y_scores[y_true == 0]),
        "Tumor score mean"   : y_scores[y_true == 1].mean(),
        "Tumor score median" : np.median(y_scores[y_true == 1]),
    }

    print(f"  EXTERNAL VALIDATION RESULTS: {name}  (GSE42568, imbalanced)")
    print(f"  Accuracy    : {res['Accuracy']:.4f}   (may be inflated by class imbalance)")
    print(f"  ROC-AUC     : {res['ROC-AUC']:.4f}")
    print(f"  Sensitivity : {res['Sensitivity']:.4f}   (Tumor recall = TP / [TP + FN])")
    print(f"  Specificity : {res['Specificity']:.4f}   (Normal recall = TN / [TN + FP])")
    print("  Classification Report:")
    print(
        classification_report(
            y_true, y_pred,
            target_names=["Normal", "Tumor"],
            zero_division=0,
        )
    )
    print("  Confusion Matrix:")
    print(f"                  Predicted Normal   Predicted Tumor")
    print(f"  Actual Normal         {TN:>5}             {FP:>5}")
    print(f"  Actual Tumor          {FN:>5}             {TP:>5}")
    print()
    print("  Raw LASSO score distribution:")
    print(f"  Normal : mean = {res['Normal score mean']:.3f}   median = {res['Normal score median']:.3f}")
    print(f"  Tumor  : mean = {res['Tumor score mean']:.3f}   median = {res['Tumor score median']:.3f}")
    return res

### Option A (primary): per-dataset standardisation of the panel probes

Each of the 15 retained probes is standardised using the **external set's own** per-probe mean and SD,
then the frozen LASSO coefficients and intercept are applied directly (`score = z · coef + intercept`).
This removes the platform / normalisation offset (MAS5 vs RMA) and is the standard approach for cross-study validation.

**No label leakage:** the mean and SD are computed over all 121 external samples without looking at the labels.

**Caveat:** centring on the external mean assumes a class mix similar to training (43/43). GSE42568 is 86 % tumour,
so its per-probe mean is pulled towards the tumour profile, which pushes tumour scores down towards the threshold.
This makes Option A conservative on sensitivity for a tumour-heavy cohort.

In [9]:
coef_series = pd.Series(model.coef_, index=scaler_features)
panel_coef  = coef_series[features].values

X_panel = X_test_df[features]
#population SD (ddof=0) to match StandardScaler
X_panel_z = (X_panel - X_panel.mean()) / X_panel.std(ddof=0)

y_scores_A = X_panel_z.values @ panel_coef + model.intercept_
results_A  = evaluate(y_test, y_scores_A, "Option A — per-dataset standardisation")

  EXTERNAL VALIDATION RESULTS: Option A — per-dataset standardisation  (GSE42568, imbalanced)
  Accuracy    : 0.7521   (may be inflated by class imbalance)
  ROC-AUC     : 0.8795
  Sensitivity : 0.7404   (Tumor recall = TP / [TP + FN])
  Specificity : 0.8235   (Normal recall = TN / [TN + FP])
  Classification Report:
              precision    recall  f1-score   support

      Normal       0.34      0.82      0.48        17
       Tumor       0.96      0.74      0.84       104

    accuracy                           0.75       121
   macro avg       0.65      0.78      0.66       121
weighted avg       0.88      0.75      0.79       121

  Confusion Matrix:
                  Predicted Normal   Predicted Tumor
  Actual Normal            14                 3
  Actual Tumor             27                77

  Raw LASSO score distribution:
  Normal : mean = 0.050   median = -0.074
  Tumor  : mean = 0.574   median = 0.561


### Option B: training scaler on log-aligned data

The original approach: the log-guarded external matrix goes through the **training** `scaler.transform()` and then `model.predict()`.
Because GSE42568 was already log2, the guard did not change anything, so this reproduces the original run.
It is kept for comparison: the remaining MAS5 vs RMA baseline difference is still baked into these scores.

In [10]:
y_scores_B = model.predict(X_test_scaled_full)
results_B  = evaluate(y_test, y_scores_B, "Option B — training scaler")

  EXTERNAL VALIDATION RESULTS: Option B — training scaler  (GSE42568, imbalanced)
  Accuracy    : 0.9752   (may be inflated by class imbalance)
  ROC-AUC     : 0.9010
  Sensitivity : 1.0000   (Tumor recall = TP / [TP + FN])
  Specificity : 0.8235   (Normal recall = TN / [TN + FP])
  Classification Report:
              precision    recall  f1-score   support

      Normal       1.00      0.82      0.90        17
       Tumor       0.97      1.00      0.99       104

    accuracy                           0.98       121
   macro avg       0.99      0.91      0.94       121
weighted avg       0.98      0.98      0.97       121

  Confusion Matrix:
                  Predicted Normal   Predicted Tumor
  Actual Normal            14                 3
  Actual Tumor              0               104

  Raw LASSO score distribution:
  Normal : mean = 0.280   median = 0.078
  Tumor  : mean = 1.115   median = 1.131


In [11]:
#side by side summary
comparison = pd.DataFrame({
    "Option A (per-dataset)": results_A,
    "Option B (training scaler)": results_B,
})
print(comparison.to_string(float_format=lambda v: f"{v:.3f}"))

#how far the external panel probes sit from the training scaler's centre (in log2 units)
panel_idx = [scaler_features.index(f) for f in features]
offset = pd.DataFrame({
    "train_mean (MAS5)": scaler.mean_[panel_idx],
    "external_mean (RMA)": X_panel.mean().values,
}, index=features)
offset["shift"] = offset["external_mean (RMA)"] - offset["train_mean (MAS5)"]
print()
print("Panel probe offset between datasets (log2 units):")
print(offset.round(2).to_string())

                     Option A (per-dataset)  Option B (training scaler)
Accuracy                              0.752                       0.975
ROC-AUC                               0.880                       0.901
Sensitivity                           0.740                       1.000
Specificity                           0.824                       0.824
TN                                   14.000                      14.000
FP                                    3.000                       3.000
FN                                   27.000                       0.000
TP                                   77.000                     104.000
Normal score mean                     0.050                       0.280
Normal score median                  -0.074                       0.078
Tumor score mean                      0.574                       1.115
Tumor score median                    0.561                       1.131

Panel probe offset between datasets (log2 units):
             

## Comparison with the original run

| | Original run | Option A (per-dataset) | Option B (training scaler) |
|---|---|---|---|
| Accuracy | 97.5 % | 75.2 % | 97.5 % |
| ROC-AUC | 0.901 | 0.880 | 0.901 |
| Sensitivity | 100 % | 74.0 % | 100 % |
| Specificity | 82.4 % | 82.4 % | 82.4 % |
| Confusion matrix (TN/FP/FN/TP) | 14/3/0/104 | 14/3/27/77 | 14/3/0/104 |
| Normal score mean / median | — | 0.050 / -0.074 | 0.280 / 0.078 |
| Tumor score mean / median | — | 0.574 / 0.561 | 1.115 / 1.131 |

**Takeaways**
- The log guard confirmed GSE42568 is already log2 (max ~16), so a missing log transform was **not** the issue; Option B is identical to the original run.
- The real mismatch is MAS5 vs RMA: the panel probes sit several log2 units away from the training scaler's centre (see offset table above). In Option B that offset inflates the scores (tumour mean ~1.1, normal mean ~0.28), so no tumour falls below the threshold. The 100 % sensitivity is at least partly this offset, not the model.
- Option A removes that offset. Ranking quality barely changes (AUC 0.88 vs 0.90), so the panel really does separate normal from tumour on an independent cohort. But at the fixed 0.5 threshold, sensitivity drops to 74 %, partly because centring on a 86 %-tumour cohort shifts tumour scores down (see caveat in Option A).
- The honest summary is the **AUC of ~0.88–0.90**. The absolute threshold does not transfer cleanly between MAS5 and RMA data, so the 97.5 % accuracy / 0 false negatives figure should not be quoted on its own.